erts/emulator/beam/jit/x86/instr_bs.cpp

```diff
@@ -1596,8 +1630,187 @@ struct BscSegment {
+                segs.back().action == BscSegment::NONE) {
+                seg.action = BscSegment::ACCUMULATE_FIRST;
+                segs.push_back(seg);
+                seg.action = BscSegment::STORE;
+                segs.push_back(seg);
+                continue;
+            }
+
+            auto prev = segs.back();
+            if (prev.flags & BSF_LITTLE) {
+                seg.action = BscSegment::ACCUMULATE_FIRST;
+                segs.push_back(seg);
+                seg.action = BscSegment::STORE;
+                segs.push_back(seg);
+                continue;
+            }
+
+            if (prev.effectiveSize + seg.effectiveSize <= 64) {
+                segs.pop_back();
+                prev.effectiveSize += seg.effectiveSize;
+                seg.action = BscSegment::ACCUMULATE;
+                segs.push_back(seg);
+                segs.push_back(prev);
+            } else {
+                seg.action = BscSegment::ACCUMULATE_FIRST;
+                segs.push_back(seg);
+                seg.action = BscSegment::STORE;
+                segs.push_back(seg);
+            }
+            break;
+        }
+        default:
+            segs.push_back(seg);
+            break;
+        }
+    }
+    return segs;
+}
+
+void BeamModuleAssembler::update_bin_state(x86::Gp bin_base,
+                                           x86::Gp bin_offset,
+                                           Sint size,
+                                           x86::Gp size_reg) {
+    const int x_reg_offset = offsetof(ErtsSchedulerRegisters, x_reg_array.d);
+    const int cur_bin_base =
+            offsetof(ErtsSchedulerRegisters, aux_regs.d.erl_bits_state) +
+            offsetof(struct erl_bits_state, erts_current_bin_);
+    const int cur_bin_offset =
+            offsetof(ErtsSchedulerRegisters, aux_regs.d.erl_bits_state) +
+            offsetof(struct erl_bits_state, erts_bin_offset_);
+
+    x86::Mem mem_bin_base =
+            x86::Mem(registers, cur_bin_base - x_reg_offset, sizeof(UWord));
+    x86::Mem mem_bin_offset =
+            x86::Mem(registers, cur_bin_offset - x_reg_offset, sizeof(UWord));
+
+    a.mov(bin_base, mem_bin_base);
+    a.mov(bin_offset, mem_bin_offset);
+
+    a.mov(ARG1, bin_offset);
+    if (size_reg.isValid()) {
+        a.add(mem_bin_offset, size_reg);
+    } else {
+        a.add(mem_bin_offset, imm(size));
+    }
+    a.shr(ARG1, imm(3));
+    a.add(ARG1, bin_base);
+}
+
+bool BeamModuleAssembler::need_mask(const ArgVal Val, Sint size) {
+    if (size == 64) {
+        return false;
+    } else if (always_small(Val)) {
+        auto [min, max] = getIntRange(Val);
+        return !(0 <= min && max >> size == 0);
+    } else {
+        return true;
+    }
+}
+
+void BeamModuleAssembler::set_zero(Sint effectiveSize) {
+    Label words = a.newLabel();
+    Label less_than_a_word = a.newLabel();
+    Sint store_unit = 1;
+
+    update_bin_state(RET, ARG2, -1, ARG3);
+
+    if (effectiveSize >= 128) {
```

Comment:
When the size is unknown, `REP STOSB` is quicker and more compact than manual zeroing on modern processors (after "Enhanced REP MOVSB support" was added in Ivy Bridge circa 2013). Specialized AVX implementations start to be slightly quicker after about two cache lines, but that feels rare enough that I don't think we need to care about it.

Reply:
Changed to use `REP STOSB` if the size is unknown or greater than 16 bytes.